docs/proposals/vip-4.md

```diff
@@ -0,0 +1,217 @@
+|                     `postSchemaRegistration` |        ✅         |         ✅         |   1️⃣    |
+|                    `preStoreVersionCreation` |        ✅         |         ✅         |    ✅     |
+|                   `postStoreVersionCreation` |        ✅         |         ✅         |    ✅     |
+|  `preStartOfStoreVersionIngestionForDaVinci` |       2️⃣        |        2️⃣        |   2️⃣    |
+| `postStartOfStoreVersionIngestionForDaVinci` |       2️⃣        |        2️⃣        |   2️⃣    |
+|          `postStoreVersionLeaderReplication` |        ✅         |         ❌         |    ❌     |
+|                        `preStoreVersionSwap` |        ✅         |         ✅         |    ✅     |
+|                       `postStoreVersionSwap` |        ✅         |         ✅         |    ✅     |
+|                            `preEndOfPushJob` |        ✅         |         ✅         |    ✅     |
+|                           `postEndOfPushJob` |        ✅         |         ✅         |    ✅     |
+
+The cells with numbers in them represent special cases, which are explained below:
+
+1️⃣ : Schema registration hooks in push jobs could only be invoked in cases where auto-registration is enabled and the 
+new schema originates from the push job itself, whereas schema registrations which are performed directly on the 
+controller could not trigger the hook.
+
+2️⃣ : The hook for the start of ingestion for Da Vinci Clients is tricky for a few reasons. The start of ingestion is 
+controlled by updating the Meta Store, which is a non-replicated system store updated by child controllers, so those
+must be involved (either by running the hook there in the first place, or by having some mechanism that enables the
+parent or VPJ to inform the child controllers of when the system store is eligible for getting updated, such as by
+adding a new field to the `AddVersion` admin channel command). However, see Rollback Support below for why running hooks
+in the child controller may be insufficient.
+
+### Rollback Support
+
+In order to support the ability for the hooks which return the `StoreVersionLifecycleEventOutcome` to rollback, the most
+natural way to achieve this is likely to involve the parent controller, either by having those hooks run there in the 
+first place, or by having a propagation mechanism to it:
+
+* If the hooks are executed in the child controller, the propagation mechanism might be to extend the job status check 
+  which the parent controller does periodically in order for the child to inform the parent of the need to rollback, or 
+  else build a new mechanism for the child to directly interact with the parent (or even with other child controllers 
+  directly...).
+
+* If the hooks are executed in VPJ, then it would need to interact with the parent via the controller client to trigger 
+  the rollback.
+
+### Configs
+
+There needs to be new configs:
```

Comment:
Regarding these hooks implementation, I am wondering how will they look like. 
Will it be some classes to be instantiated at run time using reflection and executed? If so, I can see it is going to be registered at LI Miltiproduct level (Like VPJ or Venice-backend), just as how partitioner implementation works. If so, it will not be very flexible for partners to register and evolve hooks implementation for their runs, as it requires new version roll out and deployment... 

For the beginning it might be fine, but I am wondering just like our custom partitioner vision, should we create a service that allows user to dynamically register hooks implementation.

Reply:
Yeah so... I think your point is totally right regarding custom partitioners. That stuff is too difficult for users to register and evolve, and dynamic registration would be super.

Although I have not made that clear in the doc (but I should...), in the case of these lifecycle hooks, the targeted stakeholder is the Venice operator, not the Venice user, so I don't think dynamic registration is critical. Dynamic registration is not completely useless in the context of hooks, but I would say it's a fairly minor nice to have. As such, I would prefer leaving it out of scope of this work. Though if in the future we end up building dynamic registration for the sake of custom partitioner, then we could certainly use it for hooks as well.